File: wheel/src/main/java/dev/yong/wheel/http/OkHttpHelper.java
package dev.yong.wheel.http;

import java.lang.reflect.Type;

import dev.yong.wheel.utils.JSON;
import okhttp3.Call;
import okhttp3.OkHttpClient;

/**
 * @author coderyong
 */
public class OkHttpHelper {

    private OkHttpClient httpClient;
    private OkHttpClient.Builder clientBuilder;
    private ParserFactory parserFactory;

    private OkHttpHelper() {
    }

    private static class OkHttpHelperHolder {
        private final static OkHttpHelper INSTANCE = new OkHttpHelper();
    }

    public static OkHttpHelper getInstance() {
        return OkHttpHelperHolder.INSTANCE;
    }

    /**
     * 创建OkHttp构建对象
     * <P>网络配置使用</P>
     *
     * @return OkHttpClient.Builder
     */
    public static OkHttpClient.Builder okHttp() {
        if (getInstance().clientBuilder == null) {
            getInstance().clientBuilder = new OkHttpClient.Builder();
        }
        return getInstance().clientBuilder;
    }

    /**
     * 设置解析工厂
     *
     * @param parserFactory ParserFactory
     */
    public static void setParserFactory(ParserFactory parserFactory) {
        getInstance().parserFactory = parserFactory;
    }

    /**
     * 获取接收工厂
     *
     * @return ParserFactory
     */
    static ParserFactory parserFactory() {
        if (getInstance().parserFactory == null) {
            getInstance().parserFactory = new ParserFactory() {
                @Override
                public <T> T parser(String content, Type type) {
                    if (content == null) {
                        return null;
                    }
                    return JSON.fromJson(content, type);
                }
            };
        }
        return getInstance().parserFactory;
    }

    /**
     * 创建OkHttpClient
     *
     * @return OkHttpClient
     */
    static OkHttpClient client() {
        if (getInstance().httpClient == null) {
            getInstance().httpClient = okHttp().build();
        }
        return getInstance().httpClient;
    }

    /**
     * 数据请求
     *
     * @param url 请求地址
     * @return RequestBuilder
     */
    public static RequestBuilder request(String url) {
        return RequestBuilder.request(url);
    }

    /**
     * 数据提交
     *
     * @param url 请求地址
     * @return RequestBuilder.SubmitBuilder
     */
    public static RequestBuilder.SubmitBuilder submit(String url) {
        return RequestBuilder.submit(url);
    }

    /**
     * 文件上传
     *
     * @param url 请求地址
     * @return RequestBuilder.MultipartBuilder
     */
    public static RequestBuilder.MultipartBuilder upload(String url) {
        return RequestBuilder.upload(url);
    }

    /**
     * 根据Tag取消请求
     *
     * @param tag
     */
    public void cancelTag(Object tag) {
        if (tag == null) return;
        for (Call call : client().dispatcher().queuedCalls()) {
            if (tag.equals(call.request().tag())) {
                call.cancel();
            }
        }
        for (Call call : client().dispatcher().runningCalls()) {
            if (tag.equals(call.request().tag())) {
                call.cancel();
            }
        }
    }

    /**
     * 取消所有请求请求
     */
    public static void cancelAll() {
        for (Call call : client().dispatcher().queuedCalls()) {
            call.cancel();
        }
        for (Call call : client().dispatcher().runningCalls()) {
            call.cancel();
        }
    }
}

File: wheel/src/main/java/dev/yong/wheel/http/OkHttpHelper.kt
@file:Suppress("unused", "UNCHECKED_CAST")

package dev.yong.wheel.http

import dev.yong.wheel.http.RequestBuilder.FormBuilder
import dev.yong.wheel.http.RequestBuilder.MultipartBuilder
import dev.yong.wheel.http.interceptor.ProgressInterceptor
import dev.yong.wheel.utils.JSON.fromJson
import okhttp3.OkHttpClient
import java.lang.reflect.Type

/**
 * @author coderyong
 */
class OkHttpHelper private constructor() {

    private var httpClient: OkHttpClient? = null
    private var clientBuilder: OkHttpClient.Builder? = null
    private var parserFactory: ParserFactory? = null

    /**
     * 进度拦截器
     */
    private var progressListener = ProgressInterceptor()

    private object OkHttpHelperHolder {
        val INSTANCE = OkHttpHelper()
    }

    /**
     * 根据Tag取消请求
     *
     * @param tag
     */
    fun cancelTag(tag: Any) {
        for (call in client().dispatcher.queuedCalls()) {
            if (tag == call.request().tag()) {
                call.cancel()
            }
        }
        for (call in client().dispatcher.runningCalls()) {
            if (tag == call.request().tag()) {
                call.cancel()
            }
        }
    }

    companion object {

        @JvmStatic
        fun getInstance(): OkHttpHelper {
            return OkHttpHelperHolder.INSTANCE
        }

        /**
         * 创建OkHttp构建对象
         * <P>网络配置使用</P>
         *
         * @return OkHttpClient.Builder
         */
        @JvmStatic
        fun okHttp(): OkHttpClient.Builder {
            val instance = getInstance()
            if (instance.clientBuilder == null) {
                instance.clientBuilder = OkHttpClient.Builder()
                instance.clientBuilder!!.addInterceptor(instance.progressListener)
            }
            return instance.clientBuilder!!
        }

        /**
         * 设置解析工厂
         *
         * @param parserFactory ParserFactory
         */
        @JvmStatic
        fun setParserFactory(parserFactory: ParserFactory) {
            getInstance().parserFactory = parserFactory
        }

        /**
         * 获取接收工厂
         *
         * @return ParserFactory
         */
        @JvmStatic
        fun parserFactory(): ParserFactory {
            val instance = getInstance()
            if (instance.parserFactory == null) {
                instance.parserFactory = object : ParserFactory {
                    override fun <T> parser(content: String, type: Type): T {
                        return fromJson(content, type)
                    }
                }
            }
            return instance.parserFactory!!
        }

        /**
         * 获取进度拦截器
         */
        @JvmStatic
        fun progressListener(): ProgressInterceptor {
            return getInstance().progressListener
        }

        /**
         * 创建OkHttpClient
         *
         * @return OkHttpClient
         */
        @JvmStatic
        fun client(): OkHttpClient {
            val instance = getInstance()
            if (instance.httpClient == null) {
                instance.httpClient = okHttp().build()
            }
            return instance.httpClient!!
        }

        /**
         * GET请求
         *
         * @param url 请求地址
         * @return RequestBuilder
         */
        @JvmStatic
        fun get(url: String): RequestBuilder {
            return RequestBuilder.get(url)
        }

        /**
         * HEAD请求
         *
         * @param url 请求地址
         * @return RequestBuilder
         */
        @JvmStatic
        fun head(url: String): RequestBuilder {
            return RequestBuilder.head(url)
        }

        /**
         * POST请求
         *
         * @param url 请求地址
         * @return FormBuilder
         */
        @JvmStatic
        fun post(url: String): FormBuilder {
            return RequestBuilder.post(url)
        }

        /**
         * DELETE请求
         *
         * @param url 请求地址
         * @return FormBuilder
         */
        @JvmStatic
        fun delete(url: String): FormBuilder {
            return RequestBuilder.delete(url)
        }

        /**
         * PUT请求
         *
         * @param url 请求地址
         * @return FormBuilder
         */
        @JvmStatic
        fun put(url: String): FormBuilder {
            return RequestBuilder.put(url)
        }

        /**
         * PATCH请求
         *
         * @param url 请求地址
         * @return FormBuilder
         */
        @JvmStatic
        fun patch(url: String): FormBuilder {
            return RequestBuilder.patch(url)
        }

        /**
         * 文件上传
         *
         * @param url 请求地址
         * @return MultipartBuilder
         */
        @JvmStatic
        fun upload(url: String): MultipartBuilder {
            return RequestBuilder.upload(url)
        }

        /**
         * 取消所有请求请求
         */
        @JvmStatic
        fun cancelAll() {
            for (call in client().dispatcher.queuedCalls()) {
                call.cancel()
            }
            for (call in client().dispatcher.runningCalls()) {
                call.cancel()
            }
        }
    }
}